sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,567 @@
+        }
+    );
+
+    Ok(())
+}
+
+/// Handles the `mark-bootstrap` subcommand
+async fn mark_bootstrap(args: Args) -> Result<()> {
+    let name = args
+        .container_id
+        .unwrap_or_else(|| usage_msg("Missing parameter 'container-id'"));
+    let mode = args
+        .mode
+        .unwrap_or_else(|| usage_msg("Missing parameter 'mode'"));
+
+    info!("Mode for '{}' is '{}'", name, mode);
+
+    // When 'mode' is 'once', the containers is marked as 'off' once it
+    // finishes. This guarantees that the the container is only started in
+    // the first boot.
+    if mode != "always" {
+        let key = Key::new(
+            KeyType::Data,
+            format!("settings.bootstrap-containers.{}.mode", name),
+        )
+        .expect("Adding container name to key resulted in invalid format");
+        let value = serialize_scalar(&"off".to_string()).context(error::Serialize)?;
+
+        let mut map = HashMap::new();
+        map.insert(key, value);
+        let settings: model::Settings = datastore::deserialization::from_map(&map)
+            .context(error::SettingsDeserialize { settings: map })?;
+
+        info!("Turning off container '{}'", name);
+        apiclient::set::set(args.socket_path, &settings)
+            .await
+            .context(error::Set)?;
+    }
+
+    Ok(())
+}
+
+async fn run() -> Result<()> {
+    let args = parse_args(env::args());
+
+    // SimpleLogger will send errors to stderr and anything less to stdout.
+    SimpleLogger::init(args.log_level, LogConfig::default()).context(error::Logger)?;
+
+    info!("bootstrap-containers started");
+
+    match args.subcommand.as_ref() {
+        "create-containers" => create_containers(args.socket_path).await?,
+        "mark-bootstrap" => mark_bootstrap(args).await?,
+        _ => usage_msg(format!("Unknown subcommand '{}'", args.subcommand)),
+    }
+
+    Ok(())
+}
+
+/// Store the args we receive on the command line
+struct Args {
+    log_level: LevelFilter,
+    socket_path: PathBuf,
+    subcommand: String,
+    container_id: Option<String>,
+    mode: Option<String>,
+}
+
+/// Print a usage message in the event a bad arg is passed
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {} SUBCOMMAND [ ARGUMENTS... ]
+
+    Subcommands:
+        create-containers
+        mark-bootstrap
+
+    Global arguments:
+        [ --socket-path PATH ]
+        [ --log-level trace|debug|info|warn|error ]
+
+    Mark bootstrap arguments:
+        --container-id CONTAINER-ID
+        --mode MODE
+
+    Socket path defaults to {}",
+        program_name, DEFAULT_API_SOCKET,
+    );
+    process::exit(2);
+}
+
+/// Prints a more specific message before exiting through usage().
+fn usage_msg<S: AsRef<str>>(msg: S) -> ! {
+    eprintln!("{}\n", msg.as_ref());
+    usage();
+}
+
+/// Parse the args to the program and return an Args struct
+fn parse_args(args: env::Args) -> Args {
```

Comment:
Can you change the arg parsing pattern to what we have in `shibaken` and `static-pods`? The reason is that the 'old' pattern of using `std::process:exit` makes argument parsing inherently un-testable. The new pattern (using a result) is testable.

```suggestion
fn parse_args(args: env::Args) -> Result<Args> {
```

Reply:
I changed it to return `Result<Stuff>`